### PR TITLE
S5128: "Bean Validation" (JSR 380) should be properly configured

### PR DESCRIPTION
**Pull request by bannmann**
Please ensure your pull request adheres to the following guidelines: 

- [x] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [x] Unit tests are passing and you provided a unit test for your fix
- [x] ITs should pass : To run ITs locally, checkout the [README](https://github.com/SonarSource/sonar-java/blob/master/README.md) of the project.
- [x] If there is a [Jira](http://jira.sonarsource.com/browse/SONARJAVA) ticket available, please make your commits and pull request start with the ticket number (SONARJAVA-XXXX)

### CODE REVIEW

File: java-checks/src/test/java/org/sonar/java/checks/MissingBeanValidationCheckTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+package org.sonar.java.checks;
+
+import org.junit.Test;
+import org.sonar.java.checks.verifier.JavaCheckVerifier;
+
+public class MissingBeanValidationCheckTest {
+  @Test
+  public void java8() {
+    JavaCheckVerifier.verify("src/test/files/checks/MissingBeanValidationCheckJava8.java", new MissingBeanValidationCheck(), 8);
```
Comment (m-g-sonar) [inline java-checks/src/test/java/org/sonar/java/checks/MissingBeanValidationCheckTest.java]: I don't see a reason to have different test for versions 8 of java.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      JavaSymbol.TypeJavaSymbol classSymbol = ((JavaType) variable.symbol().type()).getSymbol();
+      return classAndFieldAnnotationInstances(classSymbol);
+    }
+  }
+
+  private Stream<SymbolMetadata.AnnotationInstance> typeArgumentAnnotationInstances(VariableTree variable) {
+    return typeArgumentTypeTrees(variable).map(this::typeSymbol).flatMap(this::classAndFieldAnnotationInstances);
+  }
+
+  private JavaSymbol.TypeJavaSymbol typeSymbol(TypeTree type) {
```
[the remark this answers is not in the document]
Comment (bannmann) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: I don't understand, isn't that precisely what I already do? I call `symbolType()` on `type` (which is a `TypeTree`) to get a `JavaSymbol.TypeJavaSymbol`.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import org.sonar.plugins.java.api.semantic.SymbolMetadata;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5128")
```
[the remark this answers is not in the document]
Comment (bannmann) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: > having to manually set up exclusions

I think you misunderstood my intention, or I did not make it clear enough. Let me try again :-)

The idea is not to set up exclusions (blacklisting), but to limit the rule scope (whitelisting) using SonarQube's generic feature.

Maybe an example makes it clearer. This is how my company could set it up for our main product:

> Our architecture has several modules which are called "webapp" where transfer objects - special POJOs intended for (de)serialization - and resource classes (JAX-RS endpoints) reside. It is only in those classes that Bean Validation is used. If we fear (or experience) that S5128 results in false positives in other modules of the application, we could simply do the following:
> * In the SonarQube project, navigate to Administration, General Settings, Analysis Scope, Restrict Scope of Coding Rules, set **Rule Key Pattern** to `squid:S5128` and **File Path Pattern** to `**/webapp/**/*.java` and click **Save**.

If I understood your remark correctly, this is exactly the approach you suggested, but without adding any complexity to the rule itself.

---

However, to answer your question about adding a rule parameter, I am very skeptical about that approach:

* It would add a lot of complexity to the rule code.
* It may not be flexible enough for some users.
* Using the generic "rule parameter" UI for this could result in a bad user experience and could lead to configuration mistakes.
* But the worst part for me is that adding such logic to a rule could set a bad precedent: users could (rightly) argue that this or that controversial rule should have that feature as well.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    if (tree.is(Tree.Kind.CLASS)) {
+      checkFields((ClassTree) tree);
+    } else if (tree.is(Tree.Kind.METHOD)) {
+      checkMethod((MethodTree) tree);
+    }
+  }
+
+  private void checkFields(ClassTree tree) {
+    JavaSymbol.TypeJavaSymbol typeJavaSymbol = (JavaSymbol.TypeJavaSymbol) tree.symbol();
+    typeJavaSymbol.memberSymbols().stream().filter(SymbolUtils::isField).map(Symbol::declaration).map(VariableTree.class::cast)
```
[the remark this answers is not in the document]
Comment (bannmann) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Right, that whole block is gone now.

However, for the record, here's my answer: In fact, I like to have stream operations (and the method calls of any other kind of chain / fluent API) on separate lines, as well. However, the code style settings in `sonar-developer-toolset` insist on formatting it as above, and IMHO there's no point in "edit wars" with the formatter: just change its configuration accordingly, or, if you can't do that, live with its results.

**Comment by bannmann** [comment]
FYI @m-g-sonar: with or without my changes, the ruling test fails today:

```
[ERROR] org.sonar.java.it.JavaRulingTest  Time elapsed: 7.064 s  <<< ERROR!
java.lang.IllegalStateException: Fail to request versions at https://repox.sonarsource.com/api/search/versions?g=org.sonarsource.sonarqube&a=sonar-application&remote=0&repos=sonarsource-releases&v=6.7*
Caused by: com.sonar.orchestrator.http.HttpException: URL [https://repox.sonarsource.com/api/search/versions?g=org.sonarsource.sonarqube&a=sonar-application&remote=0&repos=sonarsource-releases&v=6.7*] returned code [503]
```

The build environment is the same unchanged, stateless docker container I used to run the ruling tests a few weeks ago, so a server error really seems plausible. Please have someone investigate this.
 
Also, I humbly suggest doing a post mortem on this and improving test coverage; IIRC this is the second or third time I work on a contribution but an unrelated server-side and/or build publishing problem breaks the build. No offense, though - I certainly know how hard it is to balance priorities in a corporate setting.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (m-g-sonar) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Ah, sorry if you misunderstood. My point was that `JavaSymbol.TypeJavaSymbol` is an implementation class, and I don't see any need to rely on implementation class rather than the interfaces `Symbol` and `Symbol.TypeSymbol`.

I reworked your code on my machine to check that there was no dependency over some implementation-hidden methods, and it's not the case.

So you can rewrite the method above like this, and drop this `typeSymbol(TypeTree)` method

```
private static Stream<SymbolMetadata.AnnotationInstance> typeArgumentAnnotationInstances(VariableTree variable) {
    return typeArgumentTypeTrees(variable)
      .map(TypeTree::symbolType)
      .map(Type::symbol)
      .flatMap(MissingBeanValidationCheck::classAndFieldAnnotationInstances);
  }
```

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-g-sonar) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Well, on this I don't fully agree. And maybe I haven't been clear enough as well, so I'll retry too! :smile: 

> The idea is not to set up exclusions (blacklisting), but to limit the rule scope (whitelisting) using SonarQube's generic feature.

On this I'm fully with you, I definitely prefer a whitelisting rather than a blacklisting. The thing I'm more interested in is the HOW we configure the rule. My point is: the rule is not enabled by default, the main reason being that you really need to carefully configure it to target the correct packages.
If we choose to rely on the SQ native approach, it means that we need to proceed in two steps:
1. Go through the rule list, enable the rule in your Quality Profile, remember the rule key
1. For each SonarQube project using beans, navigate to Administration, General Settings, Analysis Scope, Restrict Scope of Coding Rules, set `Rule Key Pattern` to `squid:S5128` and `File Path Pattern` to `**/webapp/**/*.java` (or whatever path depending of the project) and click Save

While it could be possible to do everything in one go:
1. enable the rule in your Quality Profile, and configure the parameter with the whitelist directly.

I know we transfer some logic to the rule, making it slightly more complex, while it's natively supported by SonarQube. But we also reduce the number of places we need to do adjustments in case of misconfiguration (only the rule, instead of all projects). Another aspect is that you will directly see the configuration of the rule while browsing rules, while the Analysis scope pattern is hidden in the settings of a project.

However, and I only realize it now (sorry), the main difference between the two approaches being that using the project analysis scope, you can re-use the same Quality Profile for all your projects without having to update the rule configuration in the QualityProfile... On this, I truly don't know what is the best option. I would personally go for the second ("_mine_"), but I'm fine going with the first ("_yours_") as well. So if I do not convince you, let it that way and I'll validate it!

> But the worst part for me is that adding such logic to a rule could set a bad precedent: users could (rightly) argue that this or that controversial rule should have that feature as well.

Well, there is a few rules already which are using this, and we don't do that much maintenance on it.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (m-g-sonar) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Yep, I'm gonna check. No reason to fight.
My config does not forces me to do that at all, so I wonder if I'm up to date now.

**Comment by m-g-sonar** [comment]
Hey @bannmann , thanks for the update and your comments.

I had a second look at your PR. Except the Implementation class being used for tests which should be move to the interface, it looks good to me. Using SonarLint, I spotted quite a lot of places where methods could be static, which made me adjust a few methods. I let you rebase and I'll then finalize the PR accordingly.

Regarding your other comments; Points taken, I'm going to:
- update our documentation
- check how we can do something to ease the work for external contribution - this is too much of a burden at the moment
- check the status of our toolset

> Also, I humbly suggest doing a post mortem on this and improving test coverage; IIRC this is the second or third time I work on a contribution but an unrelated server-side and/or build publishing problem breaks the build. No offense, though - I certainly know how hard it is to balance priorities in a corporate setting.

Yep, it's a huge pain for me as well, and I'm going to make sure it's easier in the future. Regarding the repox 404 errors, we just changed to another provider... And our documentation is not up to date. I'm going to fix this. I'll let you know when it's done.

Thanks again for your work, it's really appreciated.

Michael

**Comment by m-g-sonar** [comment]
Ok I had a look at our documentation (which does not need changes), and how/why you are trying to reach "repox.sonarsource.com".

> [ERROR] org.sonar.java.it.JavaRulingTest  Time elapsed: 7.064 s  <<< ERROR!
java.lang.IllegalStateException: Fail to request versions at https://repox.sonarsource.com/api/search/versions?g=org.sonarsource.sonarqube&a=sonar-application&remote=0&repos=sonarsource-releases&v=6.7*
Caused by: com.sonar.orchestrator.http.HttpException: URL [https://repox.sonarsource.com/api/search/versions?g=org.sonarsource.sonarqube&a=sonar-application&remote=0&repos=sonarsource-releases&v=6.7*] returned code [503]

After investigation, it seems that we are here suffering from a pretty bad timing. We just migrated our whole infrastructure from our old artifactory to a new one to store our artifacts online. Depending on when you started your PR, you may still have the old Orchestrator version in the dependencies in the SonarJava pom.xml. Our master is up-to-date, and should work outside our network (I didn't checked yet with an other account than my professional one).

In theory, rebasing your branch should solve the issue. Now, I'm not sure that the latest released version is available in maven central... But it's a matter of a few days if it's not the case. I'm sorry for the frustration it may have generated.

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bannmann) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Okay, I understand your rationale now, but I'm not convinced that we should add a whitelist to the rule itself*.

My gut feeling is that we are trying to solve a problem that people will probably not really have. In cases like this, I try to avoid speculating and instead to gather more insight before implementing things (along the lines of "YAGNI" and "minimum viable product" principles). Therefore, as the rule is already useful without its own whitelist, let's ship it and wait for feedback.

If it turns out a whitelist (or something else) is indeed needed, it should still be easy to add it. As a bonus, we will then have more insight into real use cases, and thus may end up implementing some other, better solution that the two of us of course cannot envision today.

----

<sup>* In fact, I think your ideas on where to configure whitelists are interesting and should really be pursued on the SonarQube level. Together with the observation that some rules already have them, they should feed into a discussion of potential new features / UX reorg. But that's off-topic here, of course.</sup>

**Comment by bannmann** [comment]
> Our master is up-to-date, and should work outside our network (I didn't checked yet with an other account than my professional one).

Unfortunately, it doesn't - at least not at 14a60484d58a590ed8a09c0c713bae5de20b49a1. I rebased my branch on that commit, however.

Anyway, I also pushed a new commit that addresses the remaining review comments and cleans up some variable names. Please tell me if you have new remarks or I overlooked/misunderstood any of your previous ones.

**Comment by m-g-sonar** [comment]
PR https://github.com/SonarSource/sonar-java/pull/2459 (merged now) should have fixed the build of ITs for external users... I let you give a try!

File: java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (m-g-sonar) [reply java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java]: Indeed, that's a very good point. I'm going to check the rules we have which implements such approach, and start discussing around it internally.

**Comment by m-g-sonar** [comment]
Hey.

I merged the duplicated PR (#2460), as I added on top of it the update of the metadata. Nice work!

Thanks again for your contribution. I'm going to move forward on our side on the various topics we discussed and try to make thing move in the good direction (namely: be sure that we can run ITs from outside the company, and rule configurations/whitelisting).

Cheers,
Michael

**Comment by bannmann** [comment]
> PR #2459 (merged now) should have fixed the build of ITs for external users... I let you give a try!

@m-g-sonar Yes, the build works fine now. Thanks!
